numba_dppy/utils/array_utils.py

```diff
@@ -0,0 +1,205 @@
+
+    if not isinstance(obj, np.ndarray):
+        raise TypeError(
+            "Obj is not USM backed and is not of type "
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_to_host(obj_memview)
+
+
+def as_usm_backed(obj, queue=None, usm_type="shared", copy=True):
+    """
+    Determine and return a SYCL device accesible object.
+
+    We try to determine if the provided object conforms to
+    __sycl_usm_array_interface__. If not, we create a USM memory of `usm_type`
+    and try to copy the data obj holds. If obj is not already backed by USM,
+    only numpy.ndarray is supported currently.
+
+    Args:
+        obj: Object to be tested and data copied from.
+        usm_type: USM type used in case obj is not already backed by USM.
+        queue (dpctl.SyclQueue): SYCL queue to be used in case obj is not
+            already USM backed.
+        copy (bool): Flag to determine if we copy data from obj.
+
+    Returns:
+        obj: USM backed memory.
+
+    Raises:
+        TypeError:
+            1. If obj is not already backed by USM and is not of type
+               numpy.ndarray, TypeError is raised.
+            2. If queue is not of type dpctl.SyclQueue.
+        ValueError:
+            1. In case obj is not USM backed, users need to pass
+               the SYCL queue to be used for creating new memory. ValuieError
+               is raised if queue argument is not provided.
+            2. If usm_type is not valid.
+            3. If dtype of the passed ndarray(obj) is not supported.
+    """
+    usm_mem = is_usm_backed(obj)
+
+    if queue is None:
+        raise ValueError(
+            "Queue can not be None. Please provide " "the SYCL queue to be used."
+        )
+    if not isinstance(queue, dpctl.SyclQueue):
+        raise TypeError(
+            "queue has to be of dpctl.SyclQueue type. Got %s" % (type(queue))
+        )
+
+    if usm_mem is None:
+        if not isinstance(obj, np.ndarray):
+            raise TypeError(
+                "Obj is not USM backed and is not of type "
+                "numpy.ndarray. Obj type: %s" % (type(obj))
+            )
+
+        if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+            raise ValueError(
+                "dtype is not supprted. Supported dtypes "
+                "are: %s" % (supported_numpy_dtype)
+            )
+
```

Comment:
Stride checking is needed here too.

Reply:
Can you please remind me what strides limitation is currently present in `dpctl`?

Reply:
What has the check for strides to do with `dpctl`? Can a kernel handle strided layout? What is the expected outcome if I pass a non-united strided array as an input/output argument?